src/OrchardCore/OrchardCore.DisplayManagement/Razor/RazorPage.cs

```diff
@@ -271,6 +274,52 @@ public new bool IsSectionDefined(string name)
             return zone != null && zone.Items.Count > 0;
         }
 
+        public new void DefineSection(string name, RenderAsyncDelegate section)
+        {
+            var zone = ThemeLayout.Zones[name];
+
+            var sectionShape = new Shape();
+            sectionShape.Metadata.Type = "AspSection";
+            sectionShape.Properties[nameof(RenderAsyncDelegate)] = SectionDelegate(section, sectionShape);
+
+            if (zone is Zones.ZoneOnDemand zoneOnDemand)
+            {
+                zoneOnDemand.AddAsync(sectionShape).GetAwaiter().GetResult();
+            }
+            else
+            {
+                zone.Add(sectionShape);
+            }
+        }
+
+        private RenderAsyncDelegate SectionDelegate(RenderAsyncDelegate oldDelegate, dynamic shape)
+        {
+            RenderAsyncDelegate newDelegate = async () =>
```

Comment:
Why not just returning `return async () => ...`

Or at least use `var newDelegate =  async () => ...` and then below add a new line before `return newDelegate`